custom/icds_reports/management/commands/migrate_person_cases_ucr.py

```diff
@@ -0,0 +1,94 @@
+            ("female_death_type", ),
+            ("referral_health_problem", ),
+            ("last_referral_date", ),
+            ("referral_reached_date", ),
+            ("hh_caste",
+                "CASE WHEN F_sc_count = 1 OR M_sc_count = 1 THEN 'sc'"
+                "     WHEN F_st_count = 1 OR M_st_count = 1 THEN 'st'"
+                "     ELSE NULL END"),
+            ("hh_minority", "CASE WHEN F_minority_count = 1 OR M_minority_count = 1 THEN 1 ELSE NULL END"),
+            ("disabled", "CASE WHEN F_disabled_count = 1 OR M_disabled_count = 1 THEN 1 ELSE NULL END"),
+            ("disability_type", "CONCAT_WS(' ', "
+                "CASE WHEN disability_type_1 = 1 THEN 1 ELSE NULL END, "
+                "CASE WHEN disability_type_2 = 1 THEN 2 ELSE NULL END, "
+                "CASE WHEN disability_type_3 = 1 THEN 3 ELSE NULL END, "
+                "CASE WHEN disability_type_4 = 1 THEN 4 ELSE NULL END, "
+                "CASE WHEN disability_type_5 = 1 THEN 5 ELSE NULL END, "
+                "CASE WHEN disability_type_6 = 1 THEN 6 ELSE NULL END "
+                ")"),
+            ("registered_status",
+                "CASE WHEN seeking_services = 1 THEN NULL"
+                "     WHEN seeking_services = 0 AND not_migrated = 1 THEN 0"
+                "     ELSE NULL END"),
+            ("migration_status", "CASE WHEN not_migrated = 0 THEN 1 ELSE NULL END"),
+            ("aadhar_date", ),
+            ("age_marriage", ),
+            ("is_pregnant",
+                "CASE WHEN pregnant_resident_count = 1 OR pregnant_migrant_count = 1 THEN 1 ELSE NULL END"),
+            ("marital_status", ),
+            # may need to remove if https://github.com/dimagi/commcare-hq/pull/22247 hasn't been merged
+            ("phone_number", ),
+        )
+        query_columns = [
+            "{} AS {}".format(col[0] if len(col) == 1 else col[1], col[0])
+            for col in columns
+        ]
+
+        blocks = SQLLocation.objects.filter(domain='icds-cas', location_type__name='block')
+        for block in blocks:
+            block_id = block.location_id
+            state = block.get_ancestor_of_type('state')
+            state_id = state.location_id
+
+            with get_cursor(ChildHealthMonthly) as cursor:
+                query_args = {'state_id': state_id, 'block_id': block_id, 'state_id_last_3': state_id[-3:]}
+                cursor.execute(
+                    "INSERT INTO \"{to_tablename}\" "
+                    "SELECT {columns} "
+                    "FROM \"{from_tablename}\" "
+                    "WHERE state_id = %(state_id)s "
+                    "AND lower(substring(state_id, '.{{3}}$'::text)) = %(state_id_last_3)s "
+                    "AND block_id = %(block_id)s".format(
+                        to_tablename=TO_TABLENAME,
+                        from_tablename=FROM_TABLENAME,
+                        columns=", \n".join(query_columns)
+                    ), query_args)
```

Comment:
I have no concept of how long this migration will take, but my guess is "very".  Do you know if this is on the order of hours or days?  Depending on the length, you might want to log which blocks have been migrated, so if the process fails partway through, you could skip already duplicated states.

Reply:
To be sure I'll have to start this on ICDS and then report back, but with that said, this is all in SQL and no data being calculated in python, so I'm pretty optimistic that it will be on the order of 1 hour. the filtering by block should be very quick as its indexed and small enough that calculations stay in memory. Also I'll be turning off pillows during this time, so there should be very little contention for write output

Reply:
cool, that makes sense to me - if it's that fast, there's no need to overprepare for the potential of failure